### PR TITLE
UT for ReconcileCOSInstance functionality

### PR DESCRIPTION
**Pull request by carmal891**
**What this PR does / why we need it**:

 UT for ReconcileCOSInstance functionality

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes # Part of https://github.com/kubernetes-sigs/cluster-api-provider-ibmcloud/issues/1818

**Special notes for your reviewer**:

/area provider/ibmcloud

1. Please confirm that if this PR changes any image versions, then that's the sole change this PR makes.

**Release note**:

```release-note

```

### CODE REVIEW

**Comment by carmal891** [comment]
Discussed and confirmed with @Karthik-K-N to omit portions that cover control flow over checkCOSBucket and createCOSBucket within the TestReconcileCOSInstance function. Instead it can be addressed as part of https://github.com/kubernetes-sigs/cluster-api-provider-ibmcloud/issues/2034

File: cloud/scope/powervs_cluster_test.go [hunk 1]
```diff
@@ -3043,3 +3047,529 @@ func TestCreateVPCSubnet(t *testing.T) {
...
+					ServiceInstance: &infrav1beta2.ResourceReference{
+						ID: ptr.To("test-serviceinstance-id"),
+					},
+				},
+			},
+		}
+
+		mockResourceController.EXPECT().GetInstanceByName(gomock.Any(), gomock.Any(), gomock.Any()).Return(&resourcecontrollerv2.ResourceInstance{
+			Name:  ptr.To("test-cos-resource-name"),
+			State: ptr.To("active"),
```
Comment (Karthik-K-N) [inline cloud/scope/powervs_cluster_test.go:3118]: not a blocking one but we can use like this as well ,( Here and at all other relevant places).
```
ptr.To(string(infrav1beta2.ServiceInstanceStateActive))
```

File: cloud/scope/powervs_cluster_test.go [hunk 2]
```diff
@@ -3043,3 +3047,529 @@ func TestCreateVPCSubnet(t *testing.T) {
...
+				},
+				Status: infrav1beta2.IBMPowerVSClusterStatus{
+					ServiceInstance: &infrav1beta2.ResourceReference{
+						ID: ptr.To("test-serviceinstance-id"),
+					},
+				},
+			},
+		}
+
+		mockResourceController.EXPECT().GetInstanceByName(gomock.Any(), gomock.Any(), gomock.Any()).Return(&resourcecontrollerv2.ResourceInstance{
```
Comment (Karthik-K-N) [inline cloud/scope/powervs_cluster_test.go:3116]: Can we also handle the case where cos instance does not exist and `GetInstanceByName` returns nil? so we endup creating cos instance?

File: cloud/scope/powervs_cluster_test.go [hunk 3]
```diff
@@ -3043,3 +3047,529 @@ func TestCreateVPCSubnet(t *testing.T) {
...
+			Name:  ptr.To("test-cos-resource-name"),
+			State: ptr.To("active"),
+			GUID:  ptr.To("test-cos-instance-guid"),
+		}, nil)
+
+		err := clusterScope.ReconcileCOSInstance()
+		g.Expect(err).ToNot(BeNil())
```
Comment (Karthik-K-N) [inline cloud/scope/powervs_cluster_test.go:3123]: If possible if we can check status as well https://github.com/Karthik-K-N/cluster-api-provider-ibmcloud/blob/987191d98e364d4bbfc39423d433b791dbb6b102/cloud/scope/powervs_cluster.go#L2171, that will be good.

File: cloud/scope/powervs_cluster_test.go [hunk 4]
```diff
@@ -3043,3 +3047,529 @@ func TestCreateVPCSubnet(t *testing.T) {
...
+		}
+
+		mockResourceController.EXPECT().CreateResourceInstance(gomock.Any()).Return(nil, nil, fmt.Errorf("error creating resource instance"))
+
+		cosResourceInstance, err := clusterScope.createCOSServiceInstance()
+		g.Expect(cosResourceInstance).To(BeNil())
+		g.Expect(err).ToNot(BeNil())
+	})
+
+	t.Run("When successfully creates resource instance", func(t *testing.T) {
```
Comment (Karthik-K-N) [inline cloud/scope/powervs_cluster_test.go:3546]:
```suggestion
	t.Run("When create resource instance is successfull", func(t *testing.T) {
```

**Review by Karthik-K-N — CHANGES_REQUESTED** [review]

File: cloud/scope/powervs_cluster_test.go (hunk 4 above)
Comment (carmal891) [reply cloud/scope/powervs_cluster_test.go]: Fixed

File: cloud/scope/powervs_cluster_test.go (hunk 3 above)
Comment (carmal891) [reply cloud/scope/powervs_cluster_test.go:3201]: Done

File: cloud/scope/powervs_cluster_test.go (hunk 2 above)
Comment (carmal891) [reply cloud/scope/powervs_cluster_test.go:3194]: Done

File: cloud/scope/powervs_cluster_test.go (hunk 1 above)
Comment (carmal891) [reply cloud/scope/powervs_cluster_test.go]: Done

File: cloud/scope/powervs_cluster_test.go [hunk 5]
```diff
@@ -3043,3 +3047,529 @@ func TestCreateVPCSubnet(t *testing.T) {
...
+				},
+				Status: infrav1beta2.IBMPowerVSClusterStatus{
+					ServiceInstance: &infrav1beta2.ResourceReference{
+						ID: ptr.To("test-serviceinstance-id"),
+					},
+				},
+			},
+		}
+
+		mockResourceController.EXPECT().GetInstanceByName(gomock.Any(), gomock.Any(), gomock.Any()).Return(nil, fmt.Errorf("Fetch failed"))
```
[the remark this answers is not in the document]
Comment (carmal891) [reply cloud/scope/powervs_cluster_test.go]: sry missed out. Done

File: cloud/scope/powervs_cluster_test.go [hunk 6]
```diff
@@ -3043,3 +3047,687 @@ func TestCreateVPCSubnet(t *testing.T) {
...
+			},
+		}
+
+		mockResourceController.EXPECT().GetInstanceByName(gomock.Any(), gomock.Any(), gomock.Any()).Return(nil, nil)
+
+		cosResourceInstance, err := clusterScope.checkCOSServiceInstance()
+		g.Expect(cosResourceInstance).To(BeNil())
+		g.Expect(err).To(BeNil())
+	})
+	t.Run("When cos service instance state is not active", func(t *testing.T) {
```
Comment (Karthik-K-N) [inline cloud/scope/powervs_cluster_test.go:3347]: whats the difference between this and next test

**Review by Karthik-K-N — COMMENTED** [review]
One qq, Otherwise LGTM, Thanks

File: cloud/scope/powervs_cluster_test.go (hunk 6 above)
Comment (carmal891) [reply cloud/scope/powervs_cluster_test.go]: apologies redundant test. Removed

File: cloud/scope/powervs_cluster_test.go [hunk 7]
```diff
@@ -3043,3 +3047,662 @@ func TestCreateVPCSubnet(t *testing.T) {
...
+		}
+
+		mockResourceController.EXPECT().GetInstanceByName(gomock.Any(), gomock.Any(), gomock.Any()).Return(nil, fmt.Errorf("error fetching instance by name"))
+
+		err = clusterScope.ReconcileCOSInstance()
+		g.Expect(err).ToNot(BeNil())
+		g.Expect(clusterScope.IBMPowerVSCluster.Status.COSInstance).To(BeNil())
+	})
+
+	t.Run("When COS service instance is not found in IBM Cloud and hence creates COS instance in cloud", func(t *testing.T) {
```
Comment (Amulyam24) [inline cloud/scope/powervs_cluster_test.go:3096]: How about the case where a COS service instance with a given name exists in cloud and `controllerCreated` is set to `false`?

File: cloud/scope/powervs_cluster_test.go [hunk 8]
```diff
@@ -3043,3 +3047,662 @@ func TestCreateVPCSubnet(t *testing.T) {
...
+			Name: ptr.To("test-resource-instance-name"),
+		}, nil, nil)
+
+		err = clusterScope.ReconcileCOSInstance()
+		g.Expect(err).ToNot(BeNil())
+		g.Expect(clusterScope.IBMPowerVSCluster.Status.COSInstance.ID).To(Equal(ptr.To("test-resource-instance-guid")))
+		g.Expect(clusterScope.IBMPowerVSCluster.Status.COSInstance.ControllerCreated).To(Equal(ptr.To(bool(true))))
+	})
+
+	t.Run("When check on whether bucket exist in service instance fails", func(t *testing.T) {
```
Comment (Amulyam24) [inline cloud/scope/powervs_cluster_test.go:3244]: can you what scenario this is covering? I'm not quite clear here

File: cloud/scope/powervs_cluster_test.go [hunk 9]
```diff
@@ -3043,3 +3047,662 @@ func TestCreateVPCSubnet(t *testing.T) {
...
+	setup := func(t *testing.T) {
+		t.Helper()
+		mockCtrl = gomock.NewController(t)
+		mockResourceController = mockRC.NewMockResourceController(mockCtrl)
+	}
+	teardown := func() {
+		mockCtrl.Finish()
+	}
+
+	t.Run("When creating cos resource instance fails due to no resource group id", func(t *testing.T) {
```
Comment (Amulyam24) [inline cloud/scope/powervs_cluster_test.go:3630]:
```suggestion
	t.Run("When creating cos resource instance fails due to missing resource group id", func(t *testing.T) {
```

File: cloud/scope/powervs_cluster_test.go [hunk 10]
```diff
@@ -3043,3 +3047,662 @@ func TestCreateVPCSubnet(t *testing.T) {
...
+						ID: ptr.To("test-serviceinstance-id"),
+					},
+				},
+			},
+		}
+
+		testScenarios := []struct {
+			name           string
+			mockError      error
+			expectedExists bool
```
Comment (Amulyam24) [inline cloud/scope/powervs_cluster_test.go:3566]:
```suggestion
			bucketExists bool
```

File: cloud/scope/powervs_cluster_test.go [hunk 11]
```diff
@@ -3043,3 +3047,662 @@ func TestCreateVPCSubnet(t *testing.T) {
...
+		}
+
+		mockResourceController.EXPECT().CreateResourceInstance(gomock.Any()).Return(nil, nil, fmt.Errorf("error creating resource instance"))
+
+		cosResourceInstance, err := clusterScope.createCOSServiceInstance()
+		g.Expect(cosResourceInstance).To(BeNil())
+		g.Expect(err).ToNot(BeNil())
+	})
+
+	t.Run("When create resource instance is successful", func(t *testing.T) {
```
Comment (Amulyam24) [inline cloud/scope/powervs_cluster_test.go:3679]:
```suggestion
	t.Run("When COS resource instance creation is successful", func(t *testing.T) {
```

File: cloud/scope/powervs_cluster_test.go [hunk 12]
```diff
@@ -3043,3 +3047,662 @@ func TestCreateVPCSubnet(t *testing.T) {
...
+	setup := func(t *testing.T) {
+		t.Helper()
+		mockCtrl = gomock.NewController(t)
+		mockResourceController = mockRC.NewMockResourceController(mockCtrl)
+		mockCOSController = mockcos.NewMockCos(mockCtrl)
+	}
+	teardown := func() {
+		mockCtrl.Finish()
+	}
+	t.Run("When checking COS bucket existence", func(t *testing.T) {
```
Comment (Amulyam24) [inline cloud/scope/powervs_cluster_test.go:3546]:
```suggestion
	t.Run("When checking if COS bucket exists in cloud", func(t *testing.T) {
```

File: cloud/scope/powervs_cluster_test.go (hunk 8 above)
Comment (carmal891) [reply cloud/scope/powervs_cluster_test.go:3244]: This covers the scenario where checkCOSBucket function returns unexpected aws error (aws error apart from s3.ErrCodeNoSuchBucket, "Forbidden" & "NotFound")

Although there is the highlighted cos client issue https://github.com/kubernetes-sigs/cluster-api-provider-ibmcloud/issues/2034 we can still add coverage over lines mentioned below. Hence added this test case covering flow over checkCOSBucket
https://github.com/kubernetes-sigs/cluster-api-provider-ibmcloud/blob/b4a2ff4d8f129d323780beb80d39a595ad81d1af/cloud/scope/powervs_cluster.go#L2221-L2224

I will update the description to 
`When checkCOSBucket fails to determine whether bucket exist in cloud due to an unexpected error
` to make it more meaningful

File: cloud/scope/powervs_cluster_test.go (hunk 7 above)
Comment (carmal891) [reply cloud/scope/powervs_cluster_test.go:3136]: Added

File: cloud/scope/powervs_cluster_test.go (hunk 9 above)
Comment (carmal891) [reply cloud/scope/powervs_cluster_test.go]: Done

File: cloud/scope/powervs_cluster_test.go (hunk 10 above)
Comment (carmal891) [reply cloud/scope/powervs_cluster_test.go]: Done

File: cloud/scope/powervs_cluster_test.go (hunk 11 above)
Comment (carmal891) [reply cloud/scope/powervs_cluster_test.go]: Done

File: cloud/scope/powervs_cluster_test.go (hunk 8 above)
Comment (Amulyam24) [reply cloud/scope/powervs_cluster_test.go]: got it, Thanks for clarifying!

**Review by Amulyam24 — APPROVED** [review]
/lgtm

**Comment by carmal891** [comment]
assign @mkumatag

**Comment by Amulyam24** [comment]
@carmal891 , please squash the commits

**Comment by carmal891** [comment]
> @carmal891 , please squash the commits

Done

**Review by Prajyot-Parab — APPROVED** [review]
/lgtm

**Comment by Karthik-K-N** [comment]
@carmal891 could you please rebase the PR.

**Comment by carmal891** [comment]
@Karthik-K-N Rebased

**Comment by carmal891** [comment]
> @carmal891 could you please rebase the PR.

Done

**Review by Prajyot-Parab — APPROVED** [review]
/lgtm

**Pull request merged**